### PR TITLE
Introduce `isProperty` to `DRI` and fix merging property and function with the same name

### PR DESCRIPTION
**Pull request by vmishenev**
Introduces `isProperty` in `DRI` to distinguish between a property and a parameterless function with the same name, e.g.,  `val f` and `fun f()`. Meanwhile, DRI allows differentiating overloads.

Fixes https://github.com/Kotlin/dokka/issues/4311, containing the corner case because a property and a function have the same `DRI` in the same source set.

Potentially, Introducing `isProperty` could also help get https://github.com/Kotlin/dokka/issues/3351 and https://github.com/Kotlin/dokka/issues/4338 fixed

### CODE REVIEW

File: dokka-subprojects/plugin-javadoc/src/test/kotlin/org/jetbrains/dokka/javadoc/JavadocAccessorNamingTest.kt [hunk 1]
```diff
@@ -80,7 +80,7 @@ internal class JavadocAccessorNamingTest : AbstractJavadocTemplateMapTest() {
 
                 assertEquals(setOf(
                     "issuesFetched",
-                    "isFoo()",
+                    "isFoo",
```
Comment (vmishenev) [inline dokka-subprojects/plugin-javadoc/src/test/kotlin/org/jetbrains/dokka/javadoc/JavadocAccessorNamingTest.kt:83]: Before the fix, it was wrong because, in this case, the getter and the backing field have the same name `isFoo`.

File: dokka-subprojects/plugin-base/src/test/kotlin/expectActuals/ExpectActualsTest.kt [hunk 2]
```diff
@@ -445,9 +446,15 @@ class ExpectActualsTest : BaseAbstractTest() {
...
-    fun `should merge an implicit-expectActual function with a single property #3685`() = testInline(
-        """
+    /**
+     * This is a corner case because a property and a function have the same DRI in the same source set
+     */
+    @Nested
+    inner class MergingPropertyAndFunction {
+
+        @Test
+        fun `should merge an implicit-expectActual function with a single property #3685`() = testInline(
```
Comment (vmishenev) [inline dokka-subprojects/plugin-base/src/test/kotlin/expectActuals/ExpectActualsTest.kt:456]: The behaviour has been changed, but this case is rare and used to cause an exception.

File: dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt [hunk 3]
```diff
@@ -87,7 +88,8 @@ public data class Callable(
     val receiver: TypeReference? = null,
     val params: List<TypeReference>,
     @property:ExperimentalDokkaApi
-    val contextParameters: List<TypeReference> = emptyList<TypeReference>()
+    val contextParameters: List<TypeReference> = emptyList(),
+    val isProperty: Boolean = false
```
Comment (whyoleg) [inline dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt:92]: Overall, the changes look fine.
The only concern I have is about invariant, that we can have non-empty `params` and `isProperty=true`.

But I don't have a good suggestion here:
1. or we can make `params` nullable (`null` meaning property) - but then `null` vs `emptyList` might be confusing a bit
2. or we can have a check in initializer, like `if(isProperty) require(params.isEmpty())`
3. or we can do nothing?

File: dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt [hunk 4]
```diff
@@ -21,7 +22,7 @@ public data class DRI(
     val extra: String? = null
 ) {
     override fun toString(): String =
-        "${packageName.orEmpty()}/${classNames.orEmpty()}/${callable?.name.orEmpty()}/${callable?.signature()
+        "${packageName.orEmpty()}/${classNames.orEmpty()}/${callable?.name.orEmpty() + if(callable?.isProperty == true) "=" else ""}/${callable?.signature()
```
Comment (whyoleg) [inline dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt]: Curious: why not embed this change into the `signature()` call?

Comment (vmishenev) [reply dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt]: because `callable?.name` is not there.
We can move it to `signature()`

Comment (whyoleg) [reply dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt]: Ah, I see, you added `=` just right after the `name`, not as part of the `signature`. Is there a reason for this?
I would expect `isProperty` to affect only `signature` output (meaning in the last part after `/`)

**Comment by whyoleg** [comment]
BTW, it would be nice to add at least one test which tests `DRI.toString` with `isProperty=true`

File: dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt (hunk 4 above)
Comment (vmishenev) [reply dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt]: `Callable. signature()` prints only a receiver and a param list and no name. I guess you would expect that the `name` also affects the `signature` output. This is probably not a good API - #3365.
For the sake of readability, the trait `isProperty` was bound to the name.
Anyway, I do not have good ideas on how to use `isProperty` in `Callable.signature()`. If you have any, I can implement one of them.

Comment (whyoleg) [reply dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt]: I would really just use `=` instead of parameters in `signature`.

Here is an example of the implementation for `signature` (uses `buildString` for readability):
```kotlin
public fun signature(): String = buildString {
    /**
     * Compatibility with [signature] without context parameters must be preserved,
     * as package-list (e.g. `DefaultExternalLocationProvider` in the base plugin) and unit tests
     * rely on `dri.toString`
     */
    @OptIn(ExperimentalDokkaApi::class)
    if (contextParameters.isNotEmpty()) {
        contextParameters.joinTo(this, separator = "#", postfix = "#")
    }
    append(receiver?.toString().orEmpty()).append("#")
    if (isProperty) {
        append("=") // `params` are replaced with `=`
    } else {
        params.joinTo(this, separator = "#")
    }
}
```

Here are some examples:
```
 com.example//function/#kotlin.Int - fun function(int: Int)
 com.example//function/#           - fun function()
 com.example//property/#=          - val property: Unit
```
(first `#` is because of the empty receiver - nothing changed here)

I feel like this way it's a bit more consistent, and we have `name` as a separate segment.
WDYT?

File: dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt [hunk 5]
```diff
@@ -87,8 +88,13 @@ public data class Callable(
     val receiver: TypeReference? = null,
     val params: List<TypeReference>,
     @property:ExperimentalDokkaApi
-    val contextParameters: List<TypeReference> = emptyList<TypeReference>()
+    val contextParameters: List<TypeReference> = emptyList(),
+    val isProperty: Boolean = false
 ) {
+    init {
+        if(isProperty) require(params.isEmpty())
```
Comment (whyoleg) [inline dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt]: minor: it might make sense to emit some message, even if it's mostly an assertion, something like `Callable $name is property, but has the following parameters: $params`

File: dokka-subprojects/plugin-javadoc/src/test/kotlin/org/jetbrains/dokka/javadoc/JavadocAccessorNamingTest.kt (hunk 1 above)
Comment (whyoleg) [reply dokka-subprojects/plugin-javadoc/src/test/kotlin/org/jetbrains/dokka/javadoc/JavadocAccessorNamingTest.kt:83]: This means that this PR also fixes the first point from #3351, right?

**Comment by whyoleg** [comment]
One last question, I promise :)

> Potentially, it could also fix https://github.com/Kotlin/dokka/issues/3351 and https://github.com/Kotlin/dokka/issues/4338

Will you add tests for those in a separate PR then?

**Comment by vmishenev** [comment]
As we decided not to break `package-list` https://github.com/Kotlin/dokka/pull/4344#discussion_r2501226860, the old version of `DRI.toString`/`Callable.toString` has been restored. 

> Will you add tests for those in a separate PR then?

Sure

File: dokka-subprojects/plugin-javadoc/src/test/kotlin/org/jetbrains/dokka/javadoc/JavadocAccessorNamingTest.kt (hunk 1 above)
Comment (vmishenev) [reply dokka-subprojects/plugin-javadoc/src/test/kotlin/org/jetbrains/dokka/javadoc/JavadocAccessorNamingTest.kt:83]: No, this test is about the Javadoc format.
To fix #3351, `JavadocExternalLocationProvider.kt` should be fixed as well. 

I updated the PR description to "Introducing isProperty could also help get https://github.com/Kotlin/dokka/issues/3351 and https://github.com/Kotlin/dokka/issues/4338 fixed" to avoid GitHub's auto-linking.

Comment (whyoleg) [reply dokka-subprojects/plugin-javadoc/src/test/kotlin/org/jetbrains/dokka/javadoc/JavadocAccessorNamingTest.kt:83]: Thanks, now I see!

File: dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt [hunk 6]
```diff
@@ -87,8 +88,16 @@ public data class Callable(
...
     val params: List<TypeReference>,
     @property:ExperimentalDokkaApi
-    val contextParameters: List<TypeReference> = emptyList<TypeReference>()
+    val contextParameters: List<TypeReference> = emptyList(),
+    val isProperty: Boolean = false
 ) {
+    init {
+        if (isProperty) require(
+            params.isEmpty(),
+            { "Callable \$name is property, but has the following parameters: $params" })
```
Comment (whyoleg) [inline dokka-subprojects/core/src/main/kotlin/org/jetbrains/dokka/links/DRI.kt]:
```suggestion
            { "Callable $name is property, but has the following parameters: $params" })
```
